### PR TITLE
List eCRs on eCR Viewer home page

### LINKED ISSUE #1779: List all eCR  on homepage
## Action Requested
All eCRs available in the backing datastore (S3 or Postgres) are available in a list view on the eCR Viewer homepage.

## Acceptance Criteria
* Homepage of viewer lists all eCRs in the bucket.
* When an ID is clicked on, you're taken to the viewer for that eCR.
* Table shows IDs and date of addition for each eCR.

Searching by ID and pagination will be covered in separate tickets!

## Testing
* Will need to test with both local installations (Postgres) and on the cloud.
* Unit and e2e tests required.

## Designs
See the [Figma prototype](https://www.figma.com/file/tXWUeV3Kjehd7VNqHojZYq/eCR-Viewer?type=design&node-id=3160-36760&mode=design&t=OrkjS29Lxcdxw1M8-0). 

## Background & Context
See the epic for more information about this body of work.

### PR DESCRIPTION
**Pull request by angelathe**
# PULL REQUEST

## Summary
**Main story:**
All eCRs available in the backing datastore (S3 or Postgres) are available in a list view on the eCR Viewer homepage.
- Add API endpoint `list-fhir-data` that querys a list of eCR IDs and date metadata (`s3` or `postgres`)
- Helper functions to process response body data from API calls to a consistent type (array of objects with eCR IDs and formatted dates
- Build home page to list eCR IDs (and stored date) in a table. Clicking on the eCR ID will redirect the user to the individual eCR viewer for that ID.
- Add snapshot test, unit tests and e2e (Cypress) tests.

**Misc:**
- Design tweaks for the home page
- Move health check API route to `src/app/api`

## Related Issue
Fixes #1779 

## Screenshots
Screenshot 2024-05-14 at 02 03 21

## Additional Info
At this stage, we are not storing date values in Postgres. Temporarily, when the data source is Postgres, the code inserts "N/A" into the date column.

## Checklist

- [ ] Pull down branch & access the homepage of the eCR Viewer ([http://localhost:3000/](http://localhost:3000/))
- [ ] Check that clicking the eCR ID links will route you to the correct individual viewer for that eCR
- [ ] Check that switching between `s3` and `postgres` as sources in `.env.local` will correctly change which eCRs are listed on the homepage.

[//]: # (PR title: Remember to name your PR descriptively!)

### CODE REVIEW

**Comment by sarahtress** [comment]
> @sarahtress Quick side question, do you have any thoughts on what the homepage should look like when there are no eCRs to load? Currently, it will only show the column names (`eCR ID` and `Stored Date`)

great question! let's ignore that for now, if possible (as in showing just the headers is fine). to me that falls into error territory (are there simply no eCRs in storage? is the connection to storage not working? etc) and it feels unlikely for there to be no eCRs in storage, more likely for there to be an error with connection to storage. but I'll make sure to make error designs for the homepage too when I work on error designs overall! thanks!

File: containers/ecr-viewer/src/app/api/route.ts
Comment (angelathe) [inline containers/ecr-viewer/src/app/api/route.ts:1]: @devs I had to move the health check route to the `/api` folder in order to add the homepage file. Should this be called anywhere?

File: containers/ecr-viewer/src/app/page.tsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+};
+
+/**
+ * Renders a list of eCR data with viewer.
+ * @param listFhirData - The list of eCRs to render.
+ * @returns The JSX element (table) representing the rendered list of eCRs.
+ */
+function renderListECRViewer(listFhirData: ListEcr): JSX.Element {
+  const header = ["eCR ID", "Stored Date"];
+  return (
```
Comment (angelathe) [inline containers/ecr-viewer/src/app/page.tsx:68]: Also wondering if the home page should also have a `main` landmark for the entire page

Comment (angelathe) [reply containers/ecr-viewer/src/app/page.tsx:68]: Added

File: containers/ecr-viewer/src/app/ListEcrViewer.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+          <tr>
+            {header.map((column) => (
+              <th key={`${column}`} scope="col">
+                {column}
+              </th>
+            ))}
+          </tr>
+        </thead>
+        <tbody>
+          {listFhirData.map((item, index) => {
```
Comment (KennethSkylight) [inline containers/ecr-viewer/src/app/ListEcrViewer.tsx:36]: Can we move this to a separate function or component?

File: containers/ecr-viewer/src/app/api/services/listEcrDataService.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  } else {
+    throw Error("Invalid Source");
+  }
+}
+
+/**
+ * Retrieves array of eCR IDs from PostgreSQL database.
+ * @returns A promise resolving to a NextResponse object.
+ */
+export const list_postgres = async () => {
```
Comment (KennethSkylight) [inline containers/ecr-viewer/src/app/api/services/listEcrDataService.ts:38]: Does this need to be exported?

File: containers/ecr-viewer/src/app/api/services/listEcrDataService.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      throw Error(error.message);
+    }
+  }
+};
+
+/**
+ * Retrieves array of eCRs and their metadata from S3.
+ * @returns A promise resolving to a NextResponse object.
+ */
+export const list_s3 = async () => {
```
Comment (KennethSkylight) [inline containers/ecr-viewer/src/app/api/services/listEcrDataService.ts:59]: same here

File: containers/ecr-viewer/src/app/services/processService.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  ecr_id: string;
+  dateModified: string;
+}[];
+
+/**
+ * Processes a list of eCR data retrieved from an S3 bucket.
+ * @param responseBody - The response body containing eCR data from S3.
+ * @returns - The processed list of eCR IDs and dates.
+ */
+export const processListS3 = (
```
Comment (KennethSkylight) [inline containers/ecr-viewer/src/app/services/processService.tsx:14]: Is there a reason why this is here instead of the listEcrDataService?

**Review by KennethSkylight — COMMENTED** [review]
Some questions

File: containers/ecr-viewer/src/app/api/services/listEcrDataService.ts (hunk 4 above)
Comment (angelathe) [reply containers/ecr-viewer/src/app/api/services/listEcrDataService.ts:59]: Done!

File: containers/ecr-viewer/src/app/api/services/listEcrDataService.ts (hunk 3 above)
Comment (angelathe) [reply containers/ecr-viewer/src/app/api/services/listEcrDataService.ts:38]: Done! Removed export

File: containers/ecr-viewer/src/app/services/processService.tsx (hunk 5 above)
Comment (angelathe) [reply containers/ecr-viewer/src/app/services/processService.tsx]: Moved!

File: containers/ecr-viewer/src/app/ListEcrViewer.tsx (hunk 2 above)
Comment (angelathe) [reply containers/ecr-viewer/src/app/ListEcrViewer.tsx]: Moved!

**Review by KennethSkylight — APPROVED** [review]

**Pull request merged**